### PR TITLE
[tellstick] Fix local bridge handler update

### LINKED ISSUE #15592: [tellstick] Existing sensors and device are added as new for every update on local bridge
## Expected Behavior
For every update using the local bridge handler sensors and devices are updated (not added), i.e. equal behaviour as the Net bridge handler.

## Current Behavior
For every refresh a new list of sensors and devices are added. This may lead to memory being exhausted and unexpected behaviour.
This due to the indexOf() function used on deviceList and sensorList always return -1 when called on existing sensors and devices.

**TelldusLocalBridgeHandler.java:**
```java
int index = this.sensorList.getSensors().indexOf(sensor);
```

## Possible Solution
Add Override for TellstickLocalSensorDTO and TellstickLocalDeviceDTO classes, similar to TellstickNetSensor and TellstickNetDevice classes:
```java
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        TellstickLocalSensorDTO other = (TellstickLocalSensorDTO) obj;
        return deviceId == other.deviceId;
    }

    @Override
    public int hashCode() {
        return deviceId;
    }
```

## Steps to Reproduce (for Bugs)
Run the Tellstick binding with the Local Bridge handler. Set to debug logging and the same thing is added again and again.

## Context
This needs to be fixed to make the Tellstick binding to be useful.

## Your Environment

* Version used: 4.0.2
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...):
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...):

### PR DESCRIPTION
**Pull request by aviborg**
Fixes #15592

### CODE REVIEW

File: bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java
Comment (lsiepel) [inline bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:121]:
```suggestion
        if (!(obj instanceof TellstickLocalDeviceDTO other)) {
```

File: bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java
Comment (lsiepel) [inline bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:127]:
```suggestion
```
depends on above.

File: bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalSensorDTO.java
Comment (lsiepel) [inline bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalSensorDTO.java:136]:
```suggestion
        if (!(obj instanceof TellstickLocalSensorDTO other)) {
```

File: bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalSensorDTO.java
Comment (lsiepel) [inline bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalSensorDTO.java:142]:
```suggestion
```
depends on above

**Review by lsiepel — CHANGES_REQUESTED** [review]
LGTM otherwise

**Review by lsiepel — APPROVED** [review]
LGTM

**Comment by aviborg** [comment]
How do you get a maintaner to approve a bugfix like this? Seems like no one cares about this binding anymore.
@kaikreuzer

**Comment by lsiepel** [comment]
> How do you get a maintaner to approve a bugfix like this? Seems like no one cares about this binding anymore. @kaikreuzer

Resources are limited, but eventually it will be picked up by a maintainer, i'm sure. There are 150+ PR's at the moment and the mayority is older then a week like this one, so please give them some time to advance.

File: bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java
Comment (maniac103) [inline bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: This check seems rather pointless, as it duplicates the instance of check above?

Comment (lsiepel) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: > This check seems rather pointless, as it duplicates the instance of check above?

There is a difference. As `instanceof` can also be a subtype, it might be a different class. Have not looked at the code to check if that might be the case in this binding. Probably not.

Comment (lolodomo) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: My feeling is that one of the 2 tests is useless, either the first or the second, no ?
I don't know what is expected by the author.

Comment (lolodomo) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: > My feeling is that one of the 2 tests is useless, either the first or the second, no ? I don't know what is expected by the author, either to have exactly same class for the two objects or not.

Comment (aviborg) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: I used the same pattern as here:
https://github.com/openhab/openhab-addons/blob/3b30d5ab11d8669e5028b3c78e42e141e0be5abd/bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/live/xml/TellstickNetSensor.java#L135
I would say that this is an example of defensive programming, i.e. try to handle any case in a predictable manner.

Comment (maniac103) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: That example doesn't have the instanceof check, though...

Comment (lsiepel) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: Instanceof and getclass are slightly different checks. If defensive programming is preferred, this is perfectly fine.
But I can’t imagine looking at the code that instanceof check passes but the getclass not. So either way would be fine and more of a personal preference regarding future fit / readability etc.

Comment (aviborg) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: > That example doesn't have the instanceof check, though...

Well, in my first commit it was a copy of that code then @lsiepel suggested the instanceof change which is made in commit 2.

Comment (lsiepel) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java:124]: It is 'free' as you did both null check and casting. So you might consider dropping the:
```if (getClass() != obj.getClass()) {```
But  as said before, it is a slightly different check and more of a personall preference. I would drop it as it is not needed (at this moment).

Comment (aviborg) [reply bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/dto/TellstickLocalDeviceDTO.java]: Ok, the class comparasion is removed in commit dc7d5e8f0e38b44a3bc93267fdda45446c04a34b

**Comment by maniac103** [comment]
That last force push added some commits which very likely are not supposed to be in this PR.

**Comment by aviborg** [comment]
Had to fix a missing signoff, going back a few commits. The rebase didn't work out as I wanted, messed up. Commit c97d2f2b8b55560619ce88586d749fb54db9fc9c is according to all historic comments.

**Review by maniac103 — APPROVED** [review]
LGTM, as far as I can tell.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Pull request merged**
